pkg/gcp/actuator/actuator.go

```diff
@@ -0,0 +1,602 @@
+func (a *Actuator) sync(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	if isGCP, err := isGCPCredentials(cr.Spec.ProviderSpec); !isGCP {
+		return err
+	}
+	logger := a.getLogger(cr)
+	logger.Debug("running sync")
+
+	infraName, err := utils.LoadInfrastructureName(a.Client, logger)
+	if err != nil {
+		return err
+	}
+
+	// First, should we update anything at all
+	needsUpdate, err := a.needsUpdate(ctx, cr)
+	if err != nil {
+		logger.WithError(err).Error("error determining whether a credentials update is needed")
+		return &actuatoriface.ActuatorError{
+			ErrReason: minterv1.CredentialsProvisionFailure,
+			Message:   fmt.Sprintf("error determining whether a credentials update is needed: %v", err),
+		}
+	}
+
+	if !needsUpdate {
+		logger.Debug("credentials already up to date")
+		return nil
+	}
+
+	cloudCredsSecret, err := a.getRootCloudCredentialsSecret(ctx, logger)
+	if err != nil {
+		logger.WithError(err).Error("issue with cloud credentials secret")
+		return err
+	}
+
+	if cloudCredsSecret.Annotations[annotatorconst.AnnotationKey] == annotatorconst.InsufficientAnnotation {
+		msg := "cloud credentials insufficient to satisfy credentials request"
+		logger.Error(msg)
+		return &actuatoriface.ActuatorError{
+			ErrReason: minterv1.InsufficientCloudCredentials,
+			Message:   msg,
+		}
+	}
+
+	if cloudCredsSecret.Annotations[annotatorconst.AnnotationKey] == annotatorconst.PassthroughAnnotation {
+		logger.Debug("provisioning with passthrough")
+		err := a.syncPassthrough(ctx, cr, cloudCredsSecret, logger)
+		if err != nil {
+			return err
+		}
+	} else if cloudCredsSecret.Annotations[annotatorconst.AnnotationKey] == annotatorconst.MintAnnotation {
+		logger.Debug("provisioning with cred minting")
+		err := a.syncMint(ctx, cr, infraName, logger)
+		if err != nil {
+			msg := "error syncing creds in mint-mode"
+			logger.WithError(err).Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   fmt.Sprintf("%s: %v", msg, err),
+			}
+		}
+	}
+
+	return nil
+}
+
+func (a *Actuator) syncPassthrough(ctx context.Context, cr *minterv1.CredentialsRequest, cloudCredsSecret *corev1.Secret, logger log.FieldLogger) error {
+	return fmt.Errorf("PASSTHROUGH ON GCP NOT YET IMPLEMENTED")
+}
+
+// syncMint handles both create and update idempotently.
+func (a *Actuator) syncMint(ctx context.Context, cr *minterv1.CredentialsRequest, infraName string, logger log.FieldLogger) error {
+	gcpSpec, err := decodeProviderSpec(a.Codec, cr)
+	if err != nil {
+		return err
+	}
+
+	gcpStatus, err := decodeProviderStatus(a.Codec, cr)
+	if err != nil {
+		return err
+	}
+
+	if gcpStatus.ServiceAccountID == "" {
+		// The service account id has a max length of 30 chars
+		// split it into 12-11-5 where the resuling string becomes:
+		// <infraName chopped to 12 chars>-<crName chopped to 11 chars>-<random 5 chars>
+		svcAcctID, err := generateUniqueNameWithFieldLimits(infraName, 12, cr.Name, 11)
```

Comment:
Whats the situation for cleanup right now, I assume we're leaking resources until we work out something with installer team?

Reply:
And does deleting a CR cleanup the objects?

Reply:
deleting the CR will delete all the created objects in the cloud, but if the cluster gets nuked (like the AWS uninstall process works), then we'll have the remnant objects to deal with